File: tests/cpp/test_vts.cpp
//  ██████   ██████ ██████████  █████████  █████   █████ █████    ███████
// ░░██████ ██████ ░░███░░░░░█ ███░░░░░███░░███   ░░███ ░░███   ███░░░░░███      ███         ███
//  ░███░█████░███  ░███  █ ░ ░███    ░░░  ░███    ░███  ░███  ███     ░░███    ░███        ░███
//  ░███░░███ ░███  ░██████   ░░█████████  ░███████████  ░███ ░███      ░███ ███████████ ███████████
//  ░███ ░░░  ░███  ░███░░█    ░░░░░░░░███ ░███░░░░░███  ░███ ░███      ░███░░░░░███░░░ ░░░░░███░░░
//  ░███      ░███  ░███ ░   █ ███    ░███ ░███    ░███  ░███ ░░███     ███     ░███        ░███
//  █████     █████ ██████████░░█████████  █████   █████ █████ ░░░███████░      ░░░         ░░░
// ░░░░░     ░░░░░ ░░░░░░░░░░  ░░░░░░░░░  ░░░░░   ░░░░░ ░░░░░    ░░░░░░░
//
//
//  License:         MIT License
//                   meshio++ default license: LICENSE
//
//  Main authors:    Vicente Mataix Ferrandiz
//
//
// VTK XML StructuredGrid (roadmap §1 tier B4, v11.6.0).

// System includes
#include <cstddef>
#include <cstdint>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>

// External includes
#include <gtest/gtest.h>

// Project includes
#include "mesh_fixtures.hpp"
#include "meshioplusplus/detail/value_io.hpp"
#include "meshioplusplus/exceptions.hpp"
#include "meshioplusplus/formats/vts.hpp"
#include "meshioplusplus/operations/voxelize.hpp"

using meshioplusplus::Mesh;
using meshioplusplus::NDArray;
using meshioplusplus::read_vts;
using meshioplusplus::read_vts_metadata;
using meshioplusplus::ReadError;
using meshioplusplus::write_vts;
using meshioplusplus::WriteError;

namespace {

Mesh vts_lattice(std::int64_t N, double Origin, double Spacing) {
    return meshioplusplus::grid({{N, N, N}}, {{Origin, Origin, Origin}},
                                {{Spacing, Spacing, Spacing}});
}

}  // namespace

TEST(Vts, RoundTripsALatticeExactly) {
    for (bool binary : {false, true}) {
        Mesh m = vts_lattice(3, -0.5, 0.25);
        NDArray pd = NDArray::Uninit(meshioplusplus::DType::Float64, {m.NumPoints()});
        for (std::size_t p = 0; p < m.NumPoints(); ++p)
            pd.As<double>()[p] = static_cast<double>(p) * 0.5;
        m.AddPointData("f", std::move(pd));
        NDArray cd = NDArray::Uninit(meshioplusplus::DType::Int64, {27u});
        for (std::size_t c = 0; c < 27; ++c)
            cd.As<std::int64_t>()[c] = static_cast<std::int64_t>(c);
        std::vector<NDArray> blocks;
        blocks.push_back(std::move(cd));
        m.AddCellData("g", std::move(blocks));

        const std::string path = mt::temp_path(".vts");
        write_vts(path, m, binary, /*zlib=*/false);
        const Mesh back = read_vts(path);

        ASSERT_EQ(back.NumPoints(), m.NumPoints());
        for (std::size_t p = 0; p < m.NumPoints(); ++p)
            for (std::size_t c = 0; c < 3; ++c)
                EXPECT_DOUBLE_EQ(meshioplusplus::detail::read_double(back.Points(), p * 3 + c),
                                 meshioplusplus::detail::read_double(m.Points(), p * 3 + c));
        ASSERT_EQ(back.NumCellBlocks(), 1u);
        EXPECT_EQ(back.Cells(0).Type(), "hexahedron");
        EXPECT_EQ(back.Cells(0).NumCells(), 27u);  // 3x3x3 cells
        ASSERT_TRUE(back.HasPointData("f"));
        for (std::size_t p = 0; p < m.NumPoints(); ++p)
            EXPECT_DOUBLE_EQ(back.PointData("f").As<double>()[p], static_cast<double>(p) * 0.5);
        ASSERT_TRUE(back.HasCellData("g"));
        for (std::size_t c = 0; c < 27; ++c)
            EXPECT_EQ(back.CellData("g", 0).As<std::int64_t>()[c], static_cast<std::int64_t>(c));
        std::remove(path.c_str());
    }
}

TEST(Vts, RefusesAMeshThatIsNotALattice) {
    Mesh m = mt::tet_mesh();
    const std::string path = mt::temp_path(".vts");
    EXPECT_THROW(write_vts(path, m, false, false), WriteError);
}

TEST(Vts, MetadataAgreesWithARealRead) {
    Mesh m = vts_lattice(2, 0.0, 1.0);
    const std::string path = mt::temp_path(".vts");
    write_vts(path, m, true, false);

    const meshioplusplus::MeshMetadata meta = read_vts_metadata(path);
    const Mesh back = read_vts(path);
    EXPECT_EQ(meta.mNumPoints, back.NumPoints());
    ASSERT_EQ(meta.mCellBlocks.size(), 1u);
    EXPECT_EQ(meta.mCellBlocks[0].mType, "hexahedron");
    EXPECT_EQ(meta.mCellBlocks[0].mNumCells, back.Cells(0).NumCells());
    std::remove(path.c_str());
}

// A hand-written file with explicit, non-lattice-checked points: the reader
// never verifies uniform spacing (only the writer does, via
// detail::lattice_from_mesh), so a genuinely curved StructuredGrid -- the
// point of the format over .vti -- must still read correctly. One cell
// (WholeExtent 0-1 on every axis, 8 points), the last point moved off the
// regular grid -- something .vti's implicit Origin/Spacing could not state
// at all, and something this scope deliberately does not verify on read
// (only .vts's WRITER requires a lattice; degenerate 2-D/1-D extents are a
// documented remainder, matching .vti's own scope, so this fixture stays
// genuinely 3-D).
TEST(Vts, ReadsAHandWrittenFileWithNonUniformPoints) {
    const std::string path = mt::temp_path(".vts");
    std::ofstream os(path, std::ios::binary);
    os << "<?xml version=\"1.0\"?>\n"
          "<VTKFile type=\"StructuredGrid\" version=\"0.1\" byte_order=\"LittleEndian\">\n"
          "<StructuredGrid WholeExtent=\"0 1 0 1 0 1\">\n"
          "<Piece Extent=\"0 1 0 1 0 1\">\n"
          "<Points>\n"
          "<DataArray type=\"Float64\" Name=\"Points\" NumberOfComponents=\"3\" format=\"ascii\">\n"
          "0 0 0\n1 0 0\n0 1 0\n1 1 0\n0 0 1\n1 0 1\n0 1 1\n1.5 1.5 1.5\n"
          "</DataArray>\n"
          "</Points>\n"
          "</Piece>\n"
          "</StructuredGrid>\n"
          "</VTKFile>\n";
    os.close();

    const Mesh back = read_vts(path);
    EXPECT_EQ(back.NumPoints(), 8u);
    ASSERT_EQ(back.NumCellBlocks(), 1u);
    EXPECT_EQ(back.Cells(0).NumCells(), 1u);
    // The last point is NOT on a regular grid -- exactly what .vti could not
    // express and .vts can.
    EXPECT_DOUBLE_EQ(meshioplusplus::detail::read_double(back.Points(), 7 * 3 + 0), 1.5);
    std::remove(path.c_str());
}

File: tests/python/test_vts.py
"""VTK XML StructuredGrid (roadmap §1 tier B4, v11.6.0).

Unlike ``.vti``, the writer's dense-lattice requirement is not the reader's:
``.vts`` states connectivity implicitly (the same index formula) but points
explicitly, so a genuinely curved structured mesh reads correctly even
though it could never be written back (there is no lattice to recover).
"""

import numpy as np
import pytest

import meshioplusplus
from meshioplusplus._grid import lattice_from_mesh
from meshioplusplus.vts import _vts


def _lattice(n=3, origin=-0.5, spacing=0.25, point_data=True, cell_data=True):
    m = meshioplusplus.grid([n, n, n], (origin,) * 3, (spacing,) * 3)
    if point_data:
        m.point_data["f"] = np.arange(len(m.points), dtype=np.float64) * 0.5
    if cell_data:
        m.cell_data["tag"] = [np.arange(n**3, dtype=np.int64)]
    return m


@pytest.mark.parametrize(
    "binary, compression", [(False, None), (True, None), (True, "zlib")]
)
def test_roundtrip(binary, compression, tmp_path):
    mesh = _lattice()
    path = tmp_path / "g.vts"
    meshioplusplus.vts.write(path, mesh, binary=binary, compression=compression)
    back = meshioplusplus.vts.read(path)
    assert np.allclose(back.points, mesh.points, atol=1e-12)
    assert len(back.cells) == 1
    assert np.array_equal(back.cells[0].data, mesh.cells[0].data)
    assert np.allclose(back.point_data["f"], mesh.point_data["f"], atol=1e-9)
    assert np.array_equal(back.cell_data["tag"][0], mesh.cell_data["tag"][0])


@pytest.mark.parametrize(
    "binary, compression", [(False, None), (True, None), (True, "zlib")]
)
def test_cross_compat(binary, compression, tmp_path):
    mesh = _lattice()
    cpp = tmp_path / "cpp.vts"
    meshioplusplus.vts.write(cpp, mesh, binary=binary, compression=compression)
    from_py = _vts.read(cpp)
    assert np.allclose(from_py.points, mesh.points, atol=1e-12)
    assert np.allclose(from_py.point_data["f"], mesh.point_data["f"], atol=1e-9)

    py = tmp_path / "py.vts"
    _vts.write(py, mesh, binary=binary, compression=compression)
    from_cpp = meshioplusplus.vts.read(py)
    assert np.allclose(from_cpp.points, mesh.points, atol=1e-12)
    assert np.allclose(from_cpp.point_data["f"], mesh.point_data["f"], atol=1e-9)
    assert np.array_equal(from_cpp.cell_data["tag"][0], mesh.cell_data["tag"][0])


def test_generic_io(tmp_path):
    mesh = _lattice(point_data=False, cell_data=False)
    path = tmp_path / "g.vts"
    meshioplusplus.write(path, mesh)
    back = meshioplusplus.read(path)
    assert np.allclose(back.points, mesh.points)
    assert meshioplusplus.sniff_format(path) == "vts"


def test_reads_a_curved_structured_grid(tmp_path):
    """The identity `.vts` exists for: points `.vti` could not state at all."""
    path = tmp_path / "curved.vts"
    path.write_text(
        '<?xml version="1.0"?>\n'
        '<VTKFile type="StructuredGrid" version="0.1" byte_order="LittleEndian">\n'
        '<StructuredGrid WholeExtent="0 1 0 1 0 1">\n'
        '<Piece Extent="0 1 0 1 0 1">\n'
        "<Points><DataArray type=\"Float64\" Name=\"Points\" "
        'NumberOfComponents="3" format="ascii">\n'
        "0 0 0\n1 0 0\n0 1 0\n1 1 0\n0 0 1\n1 0 1\n0 1 1\n1.5 1.5 1.5\n"
        "</DataArray></Points>\n"
        "</Piece></StructuredGrid></VTKFile>\n"
    )
    for reader in (meshioplusplus.vts.read, _vts.read):
        m = reader(path)
        assert len(m.points) == 8
        assert len(m.cells) == 1 and len(m.cells[0].data) == 1
        assert np.allclose(m.points[7], [1.5, 1.5, 1.5])


def test_refuses_a_mesh_that_is_not_a_lattice(tmp_path):
    path = tmp_path / "no.vts"
    tetra = meshioplusplus.Mesh(
        np.array([[0.0, 0, 0], [1, 0, 0], [0, 1, 0], [0, 0, 1]]),
        [("tetra", np.array([[0, 1, 2, 3]]))],
    )
    with pytest.raises(meshioplusplus.WriteError):
        _vts.write(path, tetra)
    surf = meshioplusplus.extract_surface(meshioplusplus.grid([2, 2, 2]))
    partial = meshioplusplus.voxelize(surf, resolution=(4, 4, 4), fill="surface")
    assert lattice_from_mesh(partial) is None
    with pytest.raises(meshioplusplus.WriteError):
        _vts.write(path, partial)


@pytest.mark.parametrize(
    "text",
    [
        '<VTKFile type="StructuredGrid"><StructuredGrid WholeExtent="0 2 0 1 0 1">'
        '<Piece Extent="0 1 0 1 0 1"/></StructuredGrid></VTKFile>',
        '<VTKFile type="UnstructuredGrid"><UnstructuredGrid/></VTKFile>',
    ],
)
def test_declines_what_it_does_not_implement(text, tmp_path):
    path = tmp_path / "bad.vts"
    path.write_text(text)
    with pytest.raises(meshioplusplus.ReadError):
        _vts.read(path)


def test_metadata_agrees_with_a_real_read(tmp_path):
    mesh = _lattice(point_data=False, cell_data=False)
    path = tmp_path / "g.vts"
    meshioplusplus.vts.write(path, mesh, binary=True)
    meta = meshioplusplus.read_metadata(path)
    back = meshioplusplus.vts.read(path)
    assert meta["num_points"] == len(back.points)
    assert meta["cell_blocks"][0]["type"] == "hexahedron"
    assert meta["cell_blocks"][0]["num_cells"] == len(back.cells[0].data)
